var/spack/repos/builtin/packages/blt/package.py

```diff
@@ -3,9 +3,90 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+import os
+import re
+
 from spack.package import *
 
 
+def spec_uses_toolchain(spec):
+    gcc_toolchain_regex = re.compile(".*gcc-toolchain.*")
+    using_toolchain = list(filter(gcc_toolchain_regex.match, spec.compiler_flags["cxxflags"]))
+    return using_toolchain
+
+
+def spec_uses_gccname(spec):
+    gcc_name_regex = re.compile(".*gcc-name.*")
+    using_gcc_name = list(filter(gcc_name_regex.match, spec.compiler_flags["cxxflags"]))
+    return using_gcc_name
+
+
+def llnl_link_helpers(options, spec, compiler):
+    # From local package:
+    if compiler.fc:
+        fortran_compilers = ["gfortran", "xlf"]
+        if any(f_comp in compiler.fc for f_comp in fortran_compilers) and (
+            "clang" in compiler.cxx
+        ):
+            # Pass fortran compiler lib as rpath to find missing libstdc++
+            libdir = os.path.join(os.path.dirname(os.path.dirname(compiler.fc)), "lib")
+            flags = ""
+            for _libpath in [libdir, libdir + "64"]:
+                if os.path.exists(_libpath):
+                    flags += " -Wl,-rpath,{0}".format(_libpath)
+            description = "Adds a missing libstdc++ rpath"
+            if flags:
+                options.append(cmake_cache_string("BLT_EXE_LINKER_FLAGS", flags, description))
+
+            excluded_root = "/usr/tce/packages/gcc/gcc-4.9.3"
```

Comment:
Why is this a completely llnl-specific path in a public spack package?

Reply:
My take on this is that the spack package should work on an LC system and at a minimum a stock rhel system of a similar version.  This specific rpathing code shouldn't be necessary, especially after the recent gcc-runtime changes, so after a merge or rebase it might be kinda moot after a test, but if we merge something like this, I'd like to at least see an issue with what needs to happen to make it portable.  In this specific case if the `gcc-runtime` package injection and rpathing doesn't fix it, which I think it will, then it would be fixed by fixing #40848 as discussed in this [thread on maintainers slack](https://spackpm.slack.com/archives/G7XED8NAD/p1706105393206209) where issues have arisen due to either not getting RPATHs to system compilers that need them or adding rpaths to system compilers with a sysroot that must not have them.  (CC: @haampie, @tgamblin, @scheibelp pulling you in from the slack thread)

Reply:
I’m fine with testing whether this is still necessary.

Reply:
Here is what happens on lassen if I remove everything in the `blt_link_helpers` function:

(this is for build with clang_14_0_5 using gcc_8_3_1 toolchain (set in the compilers flags) and cuda_11_7_0).

```
[ 13%] Linking CXX executable ../../../tests/blt_cuda_gtest_smoke
cd /dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw/blt/tests/smoke && /usr/tce/packages/cmake/cmake-3.23.1/bin/cmake -E cmake_link_script CMakeFiles/blt_cuda_gtest_smoke.dir/link.txt --verbose=1
/usr/tce/packages/clang/clang-14.0.5/bin/clang++  -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1     -Werror  -O3 -DNDEBUG -L/usr/tce/packages/cuda/cuda-11.7.0/lib64 CMakeFiles/blt_cuda_gtest_smoke.dir/blt_cuda_gtest_smoke.cpp.o -o ../../../tests/blt_cuda_gtest_smoke   -L/usr/tce/packages/cuda/cuda-11.7.0/nvidia/targets/ppc64le-linux/lib/stubs  -L/usr/tce/packages/cuda/cuda-11.7.0/nvidia/targets/ppc64le-linux/lib  -L/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3  -L/usr/tce/packages/gcc/gcc-4.9.3/lib64  -Wl,-rpath,/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/umpire-develop-xwyxixtwe7zh4eqqp7mc4spm7if7fagj/lib:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/umpire-develop-xwyxixtwe7zh4eqqp7mc4spm7if7fagj/lib64:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/camp-main-vfmvkfia3v2hm5sdtuum4tci5fbi75he/lib:/usr/tce/packages/cuda/cuda-11.7.0/lib64:/dev/shm/lassen589-1660612/clang-14.0.5.gcc.8.3.1/fmt-10.2.1-l3zyrwzbxv5ria5h6oicoytjj66cjpp6/lib64:/usr/tce/packages/gcc/gcc-4.9.3/lib64:/usr/tce/packages/clang/clang-14.0.5/release/lib ../../../lib/libgtest_main.a ../../../lib/libgtest.a -lpthread /usr/tce/packages/cuda/cuda-11.7.0/lib64/libcudart_static.a -ldl /usr/lib64/librt.so -lpthread -lcudadevrt -lcudart_static -lrt -lpthread -ldl 
../../../lib/libgtest.a(gtest-all.cc.o): In function `testing::internal::GoogleTestFailureException::GoogleTestFailureException(testing::TestPartResult const&)':
gtest-all.cc:(.text+0x22e3c): undefined reference to `std::runtime_error::runtime_error(char const*)'
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [blt/tests/smoke/CMakeFiles/blt_cuda_gtest_smoke.dir/build.make:101: tests/blt_cuda_gtest_smoke] Error 1
gmake[2]: Leaving directory '/dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw'
gmake[1]: Leaving directory '/dev/shm/lassen589-1660612/build_lassen-blueos_3_ppc64le_ib_p9-clang@14.0.5.gcc.8.3.1-xwyxixtw'
gmake[1]: *** [CMakeFiles/Makefile2:1946: blt/tests/smoke/CMakeFiles/blt_cuda_gtest_smoke.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

Things like `-L/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3` result in the wrong stdc++ being used, therefore inducing the build error.

Reply:
Please also note that Axom is already doing something similar here:
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/axom/package.py#L393

Which gives me an occasion to politely invoke @white238.

Reply:
While other helper logic for hip and cuda has been integrated to the CachedCMakePackage (discussing the topic with Spack folks internally, this was identified as the better choice), I found that this LLNL specific logic, relying on BLT, would be better defined as an helper function in the BLT package.

Reply:
Oh I see, this is a workaround for not passing the right flags somewhere.  There isn't a missing rpath, one of the cmake detection passes is finding the wrong GCC installation because the `--gcc-toolchain` flag is missing.  Without seeing your `CMakeFiles/CMakeOutput.log` file I can't say which one, but doing the same experiment with the same clang, gcc and nvcc on lassen works with this cmake configure line works just fine: `CXX=clang++ CC=clang cmake -B build -DCMAKE_C_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CUDA_FLAGS="-Xcompiler=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CXX_COMPILER=clang++ -DENABLE_CUDA=On -DCMAKE_CUDA_ARCHITECTURES=70 -DCMAKE_EXE_LINKER_FLAGS="--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" .`

I'm guessing the `CMAKE_CUDA_FLAGS` bit is missing for umpire somehow, based on the host_configs, but could be wrong.  The issue in the end is since the wrong gcc is found by cmake running either a compile or link phase for a language without the gcc-toolchain, the wrong implicit link path is added, so a wrong -L is added and it actually links with the wrong libstdc++.  Would you have an issue with adding the flags where appropriate, maybe even always in BLT based on the `spec_using_toolchain`? That way it should work everywhere that has separate gcc toolchains with separate libstdc++ instances.

Reply:
OK, it looks like the cuda flag for the toolchain is missing indeed. I can try adding that.

Regarding the rpath for the missing stdc++ with fortran, I see it in axom too. (Although it looks like the implementation is more complete there.)
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/axom/package.py#L225

Reply:
Oh interesting, I missed the fortran part, and didn't test a fortran target, maybe the rpath is separately needed for that.  The gcc-runtime package should be adding an rpath for the gcc version in use, but maybe we're not getting it because we haven't expressed to spack that we're using that gcc through clang.  @haampie is there a way to specify which gcc toolchain version is in use when not using `%gcc`?

Reply:
@trws there was indeed an oversight regarding the gcc-toolchain flags in the cuda build command. It is now fixed.
This oversight does not fix the handling of gcc-toolchain however: we have the correct flag in #39613, and locally in our builds, which verifies in the scenario described above
```
-- CUDA Compile Flags:             -Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
```
(see https://lc.llnl.gov/gitlab/radiuss/Umpire/-/jobs/1660612).

Reply:
May I suggest that since we pass --gcc-toolchain=... through the compiler flags (in compilers.yaml), we may not end up with the exact same configuration you tried ?

I am going to print out the resulting hostconfig generated by Spack, as well as the CMakeFiles/CMakeOutput.log file.

Reply:
I don't see it on the linker flags in there, which may be what's left. 
 I know this is nuts, there should be a better way to ensure this, but it will be more reliable if cmake is using the right gcc toolchain all the way down.  If that doesn't do it, go for it in the compiler flags or do what you need to do and we can break something out into a new issue to fix that part up rather than keeping this stuck.

Reply:
I agree it is not in the linker flags, I came to the same conclusion later today comparing your build line with my logs.

I am currently looking at that, that could be it. Let's follow up on Monday.

Reply:
Sounds good, sorry for my slow followup by the way, I'm out of town in all-day meetings this week so I only manage to look in sporadically. Will be back in town Monday so if you want to chat that should be a lot more feasible then.

Reply:
I see no improvement with the appropriate CMAKE_EXE_LINKER_FLAGS:

```
-- CMAKE_C_FLAGS flags are:   -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1  -Werror 
-- CMAKE_CXX_FLAGS flags are:   -Wall -Wextra  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1     -Werror 
-- CMAKE_Fortran_FLAGS flags are:  --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
-- CMAKE_EXE_LINKER_FLAGS flags are:   --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1 -Wl,-rpath,/usr/tce/packages/xl/xl-2021.09.22/lib
```

However, I now think I may have to apply the same to the CUDA linker flags:

```
-- CUDA Compile Flags:             -Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1
-- CUDA Link Flags:                -L/usr/tce/packages/cuda/cuda-11.2.0/lib64
-- BLT CUDA Separable Compilation: OFF
-- CUDA Separable Compilation:     
-- CUDA Link with NVCC:            
-- CUDA Implicit Link Libraries:   stdc++;m;gcc_s;gcc;c;gcc_s;gcc
-- CUDA Implicit Link Directories: /usr/tce/packages/cuda/cuda-11.2.0/nvidia/targets/ppc64le-linux/lib/stubs;/usr/tce/packages/cuda/cuda-11.2.0/nvidia/targets/ppc64le-linux/lib;/usr/tce/packages/gcc/gcc-4.9.3/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3;/usr/tce/packages/gcc/gcc-4.9.3/lib64;/lib64;/usr/lib64;/usr/tce/packages/clang/clang-12.0.1/release/lib;/lib;/usr/lib
```

Reply:
CMake implicitly finds paths you do not want on BlueOS. We (BLT) had to go in after it found them and strip them away because I could not find anyway to prevent it by a sane method. 

Here is how we set the variable in Axom:

https://github.com/LLNL/axom/blob/f77fd2e5f260abec7db6b7577f0482452af74e57/scripts/spack/packages/axom/package.py#L393-L410

Here is how BLT removes the paths:

https://github.com/LLNL/blt/blob/148c53ecc8bcaad5eaa4c1e39cb8144b8f1388ae/cmake/SetupCompilerOptions.cmake#L361-L375

We also have the ability to remove libraries as well:

https://github.com/LLNL/blt/blob/148c53ecc8bcaad5eaa4c1e39cb8144b8f1388ae/cmake/SetupCompilerOptions.cmake#L380-L394

Reply:
Right, I get that.  It's also kinda cool that BLT has that backported, since cmake now supports it since 3.27 it's great to have that available.

That said, I tested it and found a working solution that will work with the host toolchain in any location rather than looking for hardcoded paths a few comments up.  If we're missing one of the flags to get the `--gcc-toolchain` in all the right places then let's fix that rather than letting cmake explicitly link against the wrong library directories and hoping it happens to work at runtime with an rpath to something else.

Without being able to see the logs from one of the ones not working, I can't tell what isn't getting through.  If you want to take a look for yourself, open up `<build-dir>/CMakeFiles/CMakeOutput.log` and search for `Found candidate GCC installation`.  There will be somewhere between four and six of them, detecting the ABI and directories for C, CXX, linkers, and attendant versions for use with cuda if configured.  They should _all_ mention only the configured gcc-toolchain directory, if any of them mention the system default compiler we're missing telling it something and cmake is _explicitly forcing `-L` flags for incorrect directories_.  

After further testing, this is a refined version of the cmake invocation above I just tested with raja on lassen and rzansel with no issues, no system compiler sneaks in:
`cmake -B build -DCMAKE_C_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_CUDA_FLAGS="-Xcompiler --gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_FLAGS=--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/ -DCMAKE_EXE_LINKER_FLAGS="--gcc-toolchain=/usr/tce/packages/gcc/gcc-8.3.1/" -DCMAKE_CXX_COMPILER=clang++ DCMAKE_C_COMPILER=clang -DENABLE_CUDA=On -DCMAKE_CUDA_ARCHITECTURES=70  .`

@adrienbernede, if you can email or give me the cmake log I might be able to say what's sneaking in.  Every single flag on that line is necessary, up until the `ENABLE_CUDA`, including the C compiler since that pulls in paths too if we enable tests or certain other things.